### PR TITLE
ENH: Add FORBILD phantom in 2d

### PR DESCRIPTION
**Pull request by adler-j**
This adds the FORBILD 2d phantom to the phantoms package:

img

At some point we should perhaps consider adding the 3d case, for that I would go for simply adding a phantom generator from the format specified by the [phantomgruppe at Erlangen](http://www.imp.uni-erlangen.de/phantoms/).

### CODE REVIEW

File: odl/phantom/transmission.py [hunk 1]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+        Q = np.array([[np.cos(phi), np.sin(phi)],
+                      [-np.sin(phi), np.cos(phi)]])
+        f = phantomE[k, 5]
+        nclip = int(phantomE[k, 6])
+        equation1 = np.sum(((D.dot(Q)).dot(Vx0))**2, axis=0)
+        i = (equation1 <= 1.0)
+
+        # Handle clipping surfaces
+        if nclip > 0:
+            for j in range(1, nclip + 1):
```
Comment (adler-j) [inline odl/phantom/transmission.py]: j is not used so might as well `for _ in range(nclip):`

File: odl/phantom/transmission.py [hunk 2]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+    # Compute the phantom values in each voxel
+    image = np.zeros(space.size)
+    nclipinfo = 0
+    for k in range(phantomE.shape[0]):
+        # Handle elliptic bounds
+        Vx0 = np.array([transposeravel(xcoord) - phantomE[k, 0],
+                        transposeravel(ycoord) - phantomE[k, 1]])
+        D = np.array([[1 / phantomE[k, 2], 0],
+                      [0, 1 / phantomE[k, 3]]])
+        phi = phantomE[k, 4] * np.pi / 180
```
Comment (adler-j) [inline odl/phantom/transmission.py]: use deg2rad

File: odl/phantom/transmission.py [hunk 3]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+def _analytical_forbild_phantom(resolution, ear):
+    """Analytical description of forbild phantom.
+
+    Parameters
+    ----------
+    resolution : `bool`
+        If true, insers a small resolution insert to the left.
+    ear : `bool`
+        If true, insers a ear-like structure to the right.
+    """
```
Comment (adler-j) [inline odl/phantom/transmission.py]: define output

File: odl/phantom/transmission.py [hunk 4]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
 
+def _analytical_forbild_phantom(resolution, ear):
+    """Analytical description of forbild phantom.
+
+    Parameters
+    ----------
+    resolution : `bool`
+        If true, insers a small resolution insert to the left.
+    ear : `bool`
+        If true, insers a ear-like structure to the right.
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: insert
an ear-like structure

File: odl/phantom/transmission.py [hunk 5]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
 
 
+def _analytical_forbild_phantom(resolution, ear):
+    """Analytical description of forbild phantom.
+
+    Parameters
+    ----------
+    resolution : `bool`
+        If true, insers a small resolution insert to the left.
+    ear : `bool`
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: no backticks

File: odl/phantom/transmission.py [hunk 6]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
     return ellipse_phantom(space, ellipses)
 
 
+def _analytical_forbild_phantom(resolution, ear):
+    """Analytical description of forbild phantom.
+
+    Parameters
+    ----------
+    resolution : `bool`
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: no backticks

File: odl/phantom/transmission.py [hunk 7]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+        If true, insers a small resolution insert to the left.
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: \`\`True\`\`
insert
resolution insert - ?? resolution test pattern?

File: odl/phantom/transmission.py [hunk 8]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+        phantomC = C[:, :12]
+    elif not resolution and ear:
+        phantomE = np.vstack([E, E_cavity])
+        phantomC = C
+    elif resolution and not ear:
+        phantomE = np.vstack([leftear, E[:17, :]])
+        phantomC = C[:, :12]
+    else:
+        phantomE = np.vstack([leftear, E, E_cavity])
+        phantomC = C
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: Did you come up with all that stuff yourself? If yes, great, if no, add a reference.

File: odl/phantom/transmission.py [hunk 9]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+
+    return phantomE, phantomC
+
+
+def forbild(space, resolution=False, ear=True):
+    """Standard `FORBILD phantom` in 2 dimensions.
+
+    The forbild phantom is intended for testing CT-algorithms and is intended
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: use FORBILD throughout
CT algorithms

File: odl/phantom/transmission.py [hunk 10]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+    to be similar to a human head.
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: I don't know.. the resolution pattern doesn't look very much like anything in my head.

File: odl/phantom/transmission.py [hunk 11]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+
+    Parameters
+    ----------
+    space : `DiscreteLp`
+        The space in which the phantom should be corrected. Needs to be two-
+        dimensional.
+    resolution : `bool`, optional
+        If true, insers a small resolution insert to the left.
+    ear : `bool`, optional
+        If true, insers a ear-like structure to the right.
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: copy-paste from above - insert corrected version

File: odl/phantom/transmission.py [hunk 12]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+    --------
+    shepp_logan : A more simple phantom with similar uses, also works in 3d.
+
+    References
+    ----------
+    .. _FORBILD phantom: www.imp.uni-erlangen.de/phantoms/head/head.html
+    .. _algorithm: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3426508/
+    """
+    def transposeravel(arr):
+        """Implement matlabs ``transpose(arr(:))``."""
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: MATLAB's

File: odl/phantom/transmission.py [hunk 13]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+        Vx0 = np.array([transposeravel(xcoord) - phantomE[k, 0],
+                        transposeravel(ycoord) - phantomE[k, 1]])
+        D = np.array([[1 / phantomE[k, 2], 0],
+                      [0, 1 / phantomE[k, 3]]])
+        phi = phantomE[k, 4] * np.pi / 180
+        Q = np.array([[np.cos(phi), np.sin(phi)],
+                      [-np.sin(phi), np.cos(phi)]])
+        f = phantomE[k, 5]
+        nclip = int(phantomE[k, 6])
+        equation1 = np.sum(((D.dot(Q)).dot(Vx0))**2, axis=0)
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: pep8-ify

File: odl/phantom/transmission.py [hunk 14]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+        The space in which the phantom should be corrected. Needs to be two-
+        dimensional.
+    resolution : `bool`, optional
+        If true, insers a small resolution insert to the left.
+    ear : `bool`, optional
+        If true, insers a ear-like structure to the right.
+
+    See Also
+    --------
+    shepp_logan : A more simple phantom with similar uses, also works in 3d.
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: simpler
for similar purposes
works -> working

File: odl/phantom/transmission.py [hunk 15]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+    .. _algorithm: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3426508/
+    """
+    def transposeravel(arr):
+        """Implement matlabs ``transpose(arr(:))``."""
+        return arr.T.ravel()
+
+    if not isinstance(space, DiscreteLp):
+        raise TypeError('`space` must be a `DiscreteLp`')
+    if not space.ndim == 2:
+        raise TypeError('`space` must be two dimensional')
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: two-dimensional

File: odl/phantom/transmission.py [hunk 16]
```diff
@@ -141,6 +144,183 @@ def shepp_logan(space, modified=False):
...
+    Parameters
+    ----------
+    space : `DiscreteLp`
+        The space in which the phantom should be corrected. Needs to be two-
+        dimensional.
+    resolution : `bool`, optional
+        If true, insers a small resolution insert to the left.
+    ear : `bool`, optional
+        If true, insers a ear-like structure to the right.
+
```
Comment (kohr-h) [inline odl/phantom/transmission.py]: Output also missing here

File: odl/phantom/transmission.py (hunk 8 above)
Comment (adler-j) [inline odl/phantom/transmission.py]: I basically ported this to python

File: odl/phantom/transmission.py (hunk 10 above)
Comment (adler-j) [inline odl/phantom/transmission.py]: *similar* ;)

File: odl/phantom/transmission.py (hunk 8 above)
Comment (kohr-h) [inline odl/phantom/transmission.py]: Ah didn't see it. Okay, perhaps add it here too.

Comment (adler-j) [inline odl/phantom/transmission.py]: a bit spammy to have it twice, especially given that this is a private function.

**Comment by adler-j** [comment]
Changes fixed. Merge after CI.

**Pull request merged**